#Project Name:E-commerce Chatbot using Large Language Model (LLM)

##Overview:  
###In this capstone project, I will embark on an exciting journey to create an Industry-Specific Large Language Model (LLM) Bot using state-of-the-art pre-trained models from sources like Hugging Face. The primary objective is to build an intelligent bot that can effectively engage with users by answering questions and providing insights specific to a chosen industry. This project will not only enhance your technical skills but also provide a deep understanding of the chosen industry's nuances, challenges, and trends.

##Industry Selection::  
###Retail and E-commerce Design bots for customer service, product recommendations, and order tracking.
###I selected the Retail and E-commerce industry because:
1) Online shopping is rapidly growing

2) Businesses require automated customer support

3) chatbots are widely used in platforms like Amazon and Flipkart.



##Problem Statement:  
###In the rapidly growing retail and e-commerce industry, customers often require quick and accurate responses to their queries related to products, pricing, and services. Traditional customer support systems are time-consuming and may not always be available in real time.
###The objective of this project is to develop an AI-powered chatbot that can understand and respond to user queries related to e-commerce products. The chatbot should be capable of providing product information, answering common customer questions, and simulating real-time customer support interactions.

##Github Link:

##Let's Begin:

##1) Install Necessery libraries:

####I installed the required Python libraries needed for building and training the chatbot model in Google Colab.

In [ ]:
!pip install transformers datasets accelerate -q


In [ ]:
import transformers
import datasets
import accelerate

print("Libraries installed successfully!")

Libraries installed successfully!


##2) Import libraries:

####I imported all the required libraries and classes needed for dataset handling, model loading, and training the chatbot. The pandas library is used to read and manage the dataset, while Dataset converts the data into a format compatible with Hugging Face training. The transformers modules such as AutoTokenizer, AutoModelForCausalLM, Trainer, and TrainingArguments are used to load the pre-trained model, process text data, configure training settings, and train the chatbot model.

In [ ]:
import pandas as pd
from datasets import Dataset

from transformers import (
    AutoTokenizer,
    AutoModelForCausalLM,
    Trainer,
    TrainingArguments
)

##3) Load Dataset:

####I loaded the Flipkart e-commerce dataset into a dataframe using the read_csv() function from the pandas library. The df.head() function is used to display the first 5 rows of the dataset to verify that the data has been loaded correctly.

In [ ]:
df = pd.read_csv("/content/flipkart_ecommerce_data.csv")
df.head()

,product_name,product_category_tree,description,retail_price,discounted_price,brand
0,"Smartpro 19.5v,3.9a Replacement Charger for Va...","[""Computers >> Laptop Accessories >> Adapters ...","Key Features of Smartpro 19.5v,3.9a Replacemen...",1499.0,539.0,Smartpro
1,carrol Women's Shrug,"[""Clothing >> Women's Clothing >> Western Wear...",Key Features of carrol Women's Shrug Cotton Bl...,1800.0,599.0,carrol
2,Voylla Yellow Gold Plated Fabric Necklace,"[""Jewellery >> Necklaces & Chains >> Necklaces""]",Voylla Yellow Gold Plated Fabric Necklace\n ...,490.0,446.0,Voylla
3,Shadowfax STRAWBERRY_MAGNET Fridge Magnet,"[""Home Decor & Festive Needs >> Wall Decor & C...",Key Features of Shadowfax STRAWBERRY_MAGNET Fr...,299.0,199.0,Shadowfax
4,Lal Haveli Traditional Hand Crafted Beautiful ...,"[""Home Decor & Festive Needs >> Showpieces >> ...",Buy Lal Haveli Traditional Hand Crafted Beauti...,6500.0,4874.0,Lal Haveli


####I checked the size of the dataset using df.shape. The output Original Shape: (3000, 6) shows that the dataset contains 3000 rows and 6 columns. This helps verify the amount of data available before preprocessing and training the chatbot model.

In [ ]:
print("Original Shape:", df.shape)

Original Shape: (3000, 6)


In [ ]:
df = df.fillna("Unknown")
df.head()

,product_name,product_category_tree,description,retail_price,discounted_price,brand
0,"Smartpro 19.5v,3.9a Replacement Charger for Va...","[""Computers >> Laptop Accessories >> Adapters ...","Key Features of Smartpro 19.5v,3.9a Replacemen...",1499.0,539.0,Smartpro
1,carrol Women's Shrug,"[""Clothing >> Women's Clothing >> Western Wear...",Key Features of carrol Women's Shrug Cotton Bl...,1800.0,599.0,carrol
2,Voylla Yellow Gold Plated Fabric Necklace,"[""Jewellery >> Necklaces & Chains >> Necklaces""]",Voylla Yellow Gold Plated Fabric Necklace\n ...,490.0,446.0,Voylla
3,Shadowfax STRAWBERRY_MAGNET Fridge Magnet,"[""Home Decor & Festive Needs >> Wall Decor & C...",Key Features of Shadowfax STRAWBERRY_MAGNET Fr...,299.0,199.0,Shadowfax
4,Lal Haveli Traditional Hand Crafted Beautiful ...,"[""Home Decor & Festive Needs >> Showpieces >> ...",Buy Lal Haveli Traditional Hand Crafted Beauti...,6500.0,4874.0,Lal Haveli


##4) Create Q&A Dataset:

#### I created a custom Question & Answer dataset from the Flipkart product data to train the chatbot. A loop was used to read each product’s details such as product name, description, brand, retail price, and discounted price, and multiple question-answer pairs were generated automatically for every product. After creating the Q&A data, it was converted into a new dataframe, saved as flipkart_qa_dataset.csv, and its shape was displayed to confirm the dataset was created successfully.

In [ ]:
qa_data = []

for _, row in df.iterrows():

    product = str(row["product_name"])
    description = str(row["description"])[:250]

    brand = str(row["brand"])
    retail_price = str(row["retail_price"])
    discounted_price = str(row["discounted_price"])

    # Clean format function style answers

    # Q1: Product description
    qa_data.append({
        "text": f"Question: What is {product}?\nAnswer: {product} is a product. {description}"
    })

    # Q2: Brand
    qa_data.append({
        "text": f"Question: Which brand makes {product}?\nAnswer: The brand of this product is {brand}."
    })

    # Q3: Retail price
    qa_data.append({
        "text": f"Question: What is the retail price of {product}?\nAnswer: The retail price is ₹{retail_price}."
    })

    # Q4: Discount price
    qa_data.append({
        "text": f"Question: What is the discounted price of {product}?\nAnswer: The discounted price is ₹{discounted_price}."
    })


# Convert to dataframe
qa_df = pd.DataFrame(qa_data)

print("Q&A Dataset Shape:", qa_df.shape)

# Save dataset
qa_df.to_csv("/content/flipkart_qa_dataset.csv", index=False)

print("Q&A dataset saved successfully!")

Q&A Dataset Shape: (12000, 1)
Q&A dataset saved successfully!


####I displayed the first 20 rows of the generated Q&A dataset using qa_df.head(20). The output shows that multiple question-answer pairs were successfully created for each product, including product description, brand name, retail price, and discounted price questions. This step helps verify that the chatbot training data is properly formatted before model training.

In [ ]:
qa_df.head(20)


,text
0,"Question: What is Smartpro 19.5v,3.9a Replacem..."
1,"Question: Which brand makes Smartpro 19.5v,3.9..."
2,Question: What is the retail price of Smartpro...
3,Question: What is the discounted price of Smar...
4,Question: What is carrol Women's Shrug?\nAnswe...
5,Question: Which brand makes carrol Women's Shr...
6,Question: What is the retail price of carrol W...
7,Question: What is the discounted price of carr...
8,Question: What is Voylla Yellow Gold Plated Fa...
9,Question: Which brand makes Voylla Yellow Gold...


####I printed the first 10 question-answer entries from the generated dataset to manually inspect the chatbot training data. The loop displays each question and answer pair one by one, while "="*100 creates a separator line to make the output more readable. This helps verify that the dataset contains correctly formatted product-related questions and answers before training the model.

In [ ]:
for i in range(10):
    print(qa_df["text"][i])
    print("\n" + "="*100 + "\n")

Question: What is Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter?
Answer: Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter is a product. Key Features of Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter Designed  for Sony Vaio VGN-CR507E/J,Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter (Power Cord Included) Price: Rs. 599 


Question: Which brand makes Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter?
Answer: The brand of this product is Smartpro.


Question: What is the retail price of Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter?
Answer: The retail price is ₹1499.0.


Question: What is the discounted price of Smartpro 19.5v,3.9a Replacement Charger for Vaio Series VGN-CR507E/J 75 Adapter?
Answer: The discounted price is ₹539.0.


Question: What is carrol Women's Shrug?
Answer: carrol Women's Sh

##5)  LOAD DISTILGPT2:

####I loaded the pre-trained GPT-2 based model distilgpt2 and its tokenizer using the Hugging Face Transformers library. The tokenizer converts text into tokens that the model can understand, while AutoModelForCausalLM loads the language model for chatbot text generation. I also set tokenizer.pad_token = tokenizer.eos_token because GPT-2 models do not have a default padding token, and this prevents errors during training.

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

model_name = "distilgpt2"

tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

tokenizer.pad_token = tokenizer.eos_token

/usr/local/lib/python3.12/dist-packages/huggingface_hub/utils/_auth.py:93: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


config.json:   0%|          | 0.00/762 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/26.0 [00:00<?, ?B/s]

vocab.json: 0.00B [00:00, ?B/s]

merges.txt: 0.00B [00:00, ?B/s]

tokenizer.json: 0.00B [00:00, ?B/s]

model.safetensors:   0%|          | 0.00/353M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]

GPT2LMHeadModel LOAD REPORT from: distilgpt2
Key                                        | Status     |  | 
-------------------------------------------+------------+--+-
transformer.h.{0, 1, 2, 3, 4, 5}.attn.bias | UNEXPECTED |  | 

Notes:
- UNEXPECTED	:can be ignored when loading from different task/architecture; not ok if you expect identical arch.


generation_config.json:   0%|          | 0.00/124 [00:00<?, ?B/s]

#####I converted the Q&A dataframe into a Hugging Face dataset format using Dataset.from_pandas(qa_df). This conversion is necessary because the Hugging Face Trainer works efficiently with datasets in this format during tokenization and model training. The new dataset object now contains all chatbot question-answer pairs in a structure compatible with the training pipeline.

In [ ]:
#convert to hugging face dataset
dataset = Dataset.from_pandas(qa_df)

##6) Tokenization:

####I defined a tokenization function to prepare the dataset for training the language model. The function converts each text entry into token IDs using the tokenizer, applies padding and truncation to ensure all inputs are of equal length (64 tokens), and then creates labels by copying the input IDs for causal language modeling.

In [ ]:
def tokenize_function(example):

    tokens = tokenizer(
        example["text"],
        truncation=True,
        padding="max_length",
        max_length=64
    )

    # Labels for language modeling
    tokens["labels"] = tokens["input_ids"].copy()

    return tokens


tokenized_dataset = dataset.map(tokenize_function, batched=True)

Map:   0%|          | 0/12000 [00:00<?, ? examples/s]

##7) Training Arguments:

#####In this step, I configured and started training the chatbot model using the Hugging Face Trainer API. The TrainingArguments define how the model will be trained, including 3 epochs, batch size of 8, mixed precision (fp16=True) for faster GPU performance, and reduced logging and checkpoint saving to make training more efficient.  
#####Then, the Trainer was initialized with the model, training settings, and tokenized dataset, and training was started using trainer.train().

In [ ]:
training_args = TrainingArguments(

    output_dir="./results",

    # Faster training
    num_train_epochs=3,

    # Batch size
    per_device_train_batch_size=8,

    # Save less often
    save_steps=2000,

    # Log less often
    logging_steps=500,

     # Use mixed precision
    fp16=True,

    # Faster loading
    dataloader_num_workers=2,

    # Keep only 1 checkpoint
    save_total_limit=1,

    # Disable reports
    report_to="none"
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset
)

trainer.train()

`loss_type=None` was set in the config but it is unrecognized. Using the default loss: `ForCausalLMLoss`.


Step,Training Loss
500,1.210423
1000,0.842553
1500,0.741815
2000,0.629466
2500,0.601625
3000,0.563337
3500,0.511191
4000,0.503151
4500,0.499033


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

TrainOutput(global_step=4500, training_loss=0.6780660875108507, metrics={'train_runtime': 352.7536, 'train_samples_per_second': 102.054, 'train_steps_per_second': 12.757, 'total_flos': 587917688832000.0, 'train_loss': 0.6780660875108507, 'epoch': 3.0})

#####The output shows the training progress across 4500 steps, where the loss steadily decreases from around 1.21 to 0.49, indicating that the model is learning well. The message ForCausalLMLoss means the model is using the correct default loss function for text generation. Finally, the model was successfully saved after training completion.

##8)  SAVE MODEL:

#####I saved the trained chatbot model and tokenizer to a local directory named /content/flipkart_chatbot_model. The trainer.save_model() function stores the trained model weights, while tokenizer.save_pretrained() saves the tokenizer configuration so the same text processing can be used during inference.

In [ ]:
trainer.save_model("/content/flipkart_chatbot_model")

tokenizer.save_pretrained("/content/flipkart_chatbot_model")

print("Model saved successfully!")


Writing model shards:   0%|          | 0/1 [00:00<?, ?it/s]

Model saved successfully!


##9) Test Chabot:

#####In this step, I loaded the trained chatbot model using the Hugging Face pipeline for text generation. This allows me to quickly test the model without writing manual inference code. I passed the saved model path /content/flipkart_chatbot_model along with the tokenizer so the model can correctly process input text.

#####Then, I gave a sample question about a product brand and asked the model to generate an answer. Parameters like max_new_tokens=50 control the response length, do_sample=True enables more creative responses, and temperature=0.7 balances randomness and accuracy.

In [ ]:
from transformers import pipeline

chatbot = pipeline(
    "text-generation",
    model="/content/flipkart_chatbot_model",
    tokenizer=tokenizer
)

prompt = "Question: Which brand makes carrol Women's Shrug?\nAnswer:"

result = chatbot(
    prompt,
    max_new_tokens=30,
    do_sample=False,
    temperature=0.2,
    pad_token_id=tokenizer.eos_token_id
)

print(result[0]["generated_text"])

Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]

The following generation flags are not valid and may be ignored: ['temperature']. Set `TRANSFORMERS_VERBOSITY=info` for more details.
Passing `generation_config` together with generation-related arguments=({'temperature', 'max_new_tokens', 'pad_token_id', 'do_sample'}) is deprecated and will be removed in future versions. Please pass either a `generation_config` object OR all generation parameters explicitly, but not both.
Both `max_new_tokens` (=30) and `max_length`(=50) seem to have been set. `max_new_tokens` will take precedence. Please refer to the documentation for more information. (https://huggingface.co/docs/transformers/main/en/main_classes/text_generation)


Question: Which brand makes carrol Women's Shrug?
Answer: The brand of this product is carrol.


#####The output shows that the model correctly understood the question and generated a relevant answer: it identified the brand as Voylla, which confirms that the chatbot is working properly after training.

In [ ]:
!zip -r flipkart_chatbot_model.zip /content/flipkart_chatbot_model
from google.colab import files
files.download("flipkart_chatbot_model.zip")

  adding: content/flipkart_chatbot_model/ (stored 0%)
  adding: content/flipkart_chatbot_model/model.safetensors (deflated 7%)
  adding: content/flipkart_chatbot_model/tokenizer.json (deflated 82%)
  adding: content/flipkart_chatbot_model/training_args.bin (deflated 53%)
  adding: content/flipkart_chatbot_model/generation_config.json (deflated 25%)
  adding: content/flipkart_chatbot_model/config.json (deflated 53%)
  adding: content/flipkart_chatbot_model/tokenizer_config.json (deflated 48%)


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

##10) Gradio UI:

###To make the chatbot more user-friendly, I used Gradio to build a simple web-based interface. Gradio allows users to interact with the model by typing questions and receiving instant answers.

In [ ]:
!pip install gradio

In [ ]:
from transformers import AutoTokenizer, AutoModelForCausalLM

model_path = "/content/flipkart_chatbot_model"

tokenizer = AutoTokenizer.from_pretrained(model_path, local_files_only=True)
model = AutoModelForCausalLM.from_pretrained(model_path, local_files_only=True)

Loading weights:   0%|          | 0/76 [00:00<?, ?it/s]

In [ ]:
import os

model_directory = "/content/flipkart_chatbot_model"

if os.path.exists(model_directory):
    print(f"The model directory '{model_directory}' exists.")
else:
    print(f"The model directory '{model_directory}' does NOT exist.")

####I created a function that takes user input, converts it into tokens, and passes it to the trained model to generate a response. The output is then decoded back into readable text and returned as the chatbot’s answer.

In [ ]:
def qa_chat(user_input, history):

    prompt = f"Question: {user_input}\nAnswer:"

    inputs = tokenizer(prompt, return_tensors="pt")

    output = model.generate(
        **inputs,
        max_new_tokens=50,
        temperature=0.2,
        do_sample=False,
        pad_token_id=tokenizer.eos_token_id
    )

    response = tokenizer.decode(output[0], skip_special_tokens=True)

    answer = response.split("Answer:")[-1].strip()

    return answer

####I built the chatbot interface using Gradio. The ChatInterface connects my qa_chat function to a simple web UI where users can type questions and get answers instantly. Finally, chatbot.launch() starts the application and opens the chatbot in a browser for interaction.

In [ ]:
import gradio as gr

chatbot = gr.ChatInterface(
    fn=qa_chat,
    title="Q&A Chatbot",
    description="Ask questions about products and get answers"
)

chatbot.launch()

/usr/local/lib/python3.12/dist-packages/gradio/chat_interface.py:347: UserWarning: The 'tuples' format for chatbot messages is deprecated and will be removed in a future version of Gradio. Please set type='messages' instead, which uses openai-style 'role' and 'content' keys.
  self.chatbot = Chatbot(


It looks like you are running Gradio on a hosted Jupyter notebook, which requires `share=True`. Automatically setting `share=True` (you can turn this off by setting `share=False` in `launch()` explicitly).

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
* Running on public URL: https://75c34da69b4665d08b.gradio.live

This share link expires in 1 week. For free permanent hosting and GPU upgrades, run `gradio deploy` from the terminal in the working directory to deploy to Hugging Face Spaces (https://huggingface.co/spaces)


####Overall, using Gradio made my project more interactive, visually clear, and easy to present. It helped me focus more on the machine learning part while still providing a professional-looking chatbot interface for users.

###Conclusion:  
####In this project, I built an end-to-end e-commerce Q&A chatbot using a Flipkart product dataset. I started by loading and preprocessing the data, then converted it into structured question-answer pairs for training. After that, I tokenized the dataset and fine-tuned a pre-trained language model using Hugging Face Transformers.
####Finally, I tested the model and deployed it using Gradio to create an interactive web-based chatbot interface. The system is able to answer product-related questions such as brand, description, and pricing effectively, showing that the trained model can be used for real-world e-commerce assistance.